### PR TITLE
Add Pinned collection parameter type

### PR DESCRIPTION
**Pull request by jtschuster**
Builds off https://github.com/dotnet/runtime/pull/90057.

Adds stateful and stateless collection marshallers with GetPinnableReference methods.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IStatefulCollectionPinnableReference.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        void MethodOut(
+            [MarshalUsing(CountElementName = nameof(size))] out StatefulCollectionPinnableReference<int> pOut,
+            out int size);
+
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatefulCollectionPinnableReference<int> Return(int size);
+
+        [PreserveSig]
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatefulCollectionPinnableReference<int> ReturnPreserveSig(int size);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IStatefulCollectionPinnableReference.cs:34]: I'm fine having this if it works without a lot of ceremony. Our first version of the COM source generator should be strict matching of COM, so `HRESULT` returns. This isn't valid COM unfortunately so if we hit any issues with this scenario it is fine to drop it.

File: src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IStatelessCollectionPinnableReference.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        void MethodOut(
+            [MarshalUsing(CountElementName = nameof(size))] out StatelessCollectionPinnableReference<int> pOut,
+            out int size);
+
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatelessCollectionPinnableReference<int> Return(int size);
+
+        [PreserveSig]
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatelessCollectionPinnableReference<int> ReturnPreserveSig(int size);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IStatelessCollectionPinnableReference.cs:34]: Same as above.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
LGTM. How does this get consumed?

**Review by jkoritzinsky — APPROVED** [review]

**Comment by jtschuster** [comment]
> LGTM. How does this get consumed?

For now it just is an outline and can help make an inner loop that fails to compile if the generator generates bad code and makes it easier to inspect generated code. I'll be adding an implementation later so that it can be tested more thoroughly.

**Pull request merged**